### PR TITLE
COST-1829: Initial currency settings infrastructure support

### PR DESCRIPTION
**Pull request by dccurtis**
Thanks to @boaz0 for all the heavy lifting on this.  These changes are ported from https://github.com/project-koku/koku/pull/3070.

**Testing**
1. Create a source and hit /settings with DEVELOPMENT=True.  Verify currency options are displayed and initial value is USD
2. use Django python shell to change currency.
3. Hit /settings again and verify that initialValue for currency was changed from step 2.

***Test Results***
```
GET /api/cost-management/v1/settings/
HTTP 200 OK
Allow: GET, POST, HEAD, OPTIONS
Cache-Control: max-age=0, no-cache, no-store, must-revalidate, private
Content-Type: application/json
Expires: Fri, 03 Sep 2021 18:15:36 GMT
Vary: Accept

[
    {
        "title": "Enable tags and labels",
        "name": "api.settings.tag_managment.subform",
        "fields": [
            {
                "component": "plain-text-with-links",
                "text": "Enable your data source labels to be used as tag keys for report grouping and filtering. Changes will be reflected within 24 hours. <link>Learn more</link>",
                "linkProps": [
                    {
                        "href": "https://access.redhat.com/documentation/en-us/cost_management_service/2021/html-single/managing_cost_data_using_tagging/index#assembly-configuring-tags-and-labels-in-cost-management"
                    }
                ],
                "name": "api.settings.tag_management.form-text"
            },
            {
                "label": "OpenShift labels",
                "name": "api.settings.openshift.title",
                "variant": "h2",
                "component": "plain-text"
            },
            {
                "component": "dual-list-select",
                "name": "api.settings.tag-management.openshift.enabled",
                "options": [],
                "leftTitle": "Available labels",
                "rightTitle": "Labels for reporting",
                "noValueTitle": "No enabled tag keys",
                "noOptionsTitle": "No available tag keys",
                "filterOptionsTitle": "Filter by available tag keys",
                "filterValueTitle": "Filter by enabled tag keys",
                "filterValueText": "Remove your filter to see all enabled tag keys",
                "filterOptionsText": "Remove your filter to see all available tag keys",
                "initialValue": [],
                "clearedValue": []
            },
            {
                "label": "Amazon Web Services tags",
                "name": "api.settings.aws.title",
                "variant": "h2",
                "component": "plain-text"
            },
            {
                "component": "dual-list-select",
                "name": "api.settings.tag-management.aws.enabled",
                "options": [],
                "leftTitle": "Available tags",
                "rightTitle": "Tags for reporting",
                "noValueTitle": "No enabled tag keys",
                "noOptionsTitle": "No available tag keys",
                "filterOptionsTitle": "Filter by available tag keys",
                "filterValueTitle": "Filter by enabled tag keys",
                "filterValueText": "Remove your filter to see all enabled tag keys",
                "filterOptionsText": "Remove your filter to see all available tag keys",
                "initialValue": [],
                "clearedValue": []
            },
            {
                "label": "Azure tags",
                "name": "api.settings.azure.title",
                "variant": "h2",
                "component": "plain-text"
            },
            {
                "component": "dual-list-select",
                "name": "api.settings.tag-management.azure.enabled",
                "options": [],
                "leftTitle": "Available tags",
                "rightTitle": "Tags for reporting",
                "noValueTitle": "No enabled tag keys",
                "noOptionsTitle": "No available tag keys",
                "filterOptionsTitle": "Filter by available tag keys",
                "filterValueTitle": "Filter by enabled tag keys",
                "filterValueText": "Remove your filter to see all enabled tag keys",
                "filterOptionsText": "Remove your filter to see all available tag keys",
                "initialValue": [],
                "clearedValue": []
            },
            {
                "label": "Google Cloud Platform tags",
                "name": "api.settings.gcp.title",
                "variant": "h2",
                "component": "plain-text"
            },
            {
                "component": "dual-list-select",
                "name": "api.settings.tag-management.gcp.enabled",
                "options": [],
                "leftTitle": "Available tags",
                "rightTitle": "Tags for reporting",
                "noValueTitle": "No enabled tag keys",
                "noOptionsTitle": "No available tag keys",
                "filterOptionsTitle": "Filter by available tag keys",
                "filterValueTitle": "Filter by enabled tag keys",
                "filterValueText": "Remove your filter to see all enabled tag keys",
                "filterOptionsText": "Remove your filter to see all available tag keys",
                "initialValue": [],
                "clearedValue": []
            }
        ],
        "component": "sub-form"
    },
    {
        "title": "Currency",
        "fields": [
            {
                "label": "Select the prefered currency to view Cost Information in",
                "name": "api.settings.openshift.title",
                "component": "plain-text"
            },
            {
                "component": "select",
                "name": "api.settings.currency",
                "label": "Currency",
                "options": [
                    {
                        "value": "AUD",
                        "label": "AUD (AU$) - Australian Dollar"
                    },
                    {
                        "value": "CAD",
                        "label": "CAD (C$) - Canadian Dollar"
                    },
                    {
                        "value": "CHF",
                        "label": "CHF (CHF) - Swiss Franc"
                    },
                    {
                        "value": "CNY",
                        "label": "CNY (CN¥) - Chinese Yuan"
                    },
                    {
                        "value": "DKK",
                        "label": "DKK (Dkr) - Danish Krone"
                    },
                    {
                        "value": "EUR",
                        "label": "EUR (€) - Euro"
                    },
                    {
                        "value": "GBP",
                        "label": "GBP (£) - British Pound"
                    },
                    {
                        "value": "HKD",
                        "label": "HKD (HK$) - Hong Kong Dollar"
                    },
                    {
                        "value": "JPY",
                        "label": "JPY (¥) - Japanese Yen"
                    },
                    {
                        "value": "NOK",
                        "label": "NOK (Nkr) - Norwegian Krone"
                    },
                    {
                        "value": "NZD",
                        "label": "NZD (NZ$) - New Zealand Dollar"
                    },
                    {
                        "value": "SEK",
                        "label": "SEK (Skr) - Swedish Krona"
                    },
                    {
                        "value": "SGD",
                        "label": "SGD (S$) - Singapore Dollar"
                    },
                    {
                        "value": "USD",
                        "label": "USD ($) - United States Dollar"
                    },
                    {
                        "value": "ZAR",
                        "label": "ZAR (R) - South African Rand"
                    }
                ],
                "initialValue": "USD"
            }
        ]
    }
]

(koku_test2) docurtis@docurtis-mac koku_test2 % docker exec -it koku_server bash
(app-root) bash-4.4$ python koku/manage.py shell
Loading : /koku/.env
The .env file has been loaded.
celery broker using redis url: redis://redis:6379/1
Python 3.8.6 (default, Jan 22 2021, 11:41:28) 
[GCC 8.4.1 20200928 (Red Hat 8.4.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> from currency.common import set_currency
>>> set_currency('acct10001', currency_code='JPY')
>>> 

GET /api/cost-management/v1/settings/
HTTP 200 OK
Allow: GET, POST, HEAD, OPTIONS
Cache-Control: max-age=0, no-cache, no-store, must-revalidate, private
Content-Type: application/json
Expires: Fri, 03 Sep 2021 18:18:29 GMT
Vary: Accept

[
    {
        "title": "Enable tags and labels",
        "name": "api.settings.tag_managment.subform",
        "fields": [
            {
                "component": "plain-text-with-links",
                "text": "Enable your data source labels to be used as tag keys for report grouping and filtering. Changes will be reflected within 24 hours. <link>Learn more</link>",
                "linkProps": [
                    {
                        "href": "https://access.redhat.com/documentation/en-us/cost_management_service/2021/html-single/managing_cost_data_using_tagging/index#assembly-configuring-tags-and-labels-in-cost-management"
                    }
                ],
                "name": "api.settings.tag_management.form-text"
            },
            {
                "label": "OpenShift labels",
                "name": "api.settings.openshift.title",
                "variant": "h2",
                "component": "plain-text"
            },
            {
                "component": "dual-list-select",
                "name": "api.settings.tag-management.openshift.enabled",
                "options": [],
                "leftTitle": "Available labels",
                "rightTitle": "Labels for reporting",
                "noValueTitle": "No enabled tag keys",
                "noOptionsTitle": "No available tag keys",
                "filterOptionsTitle": "Filter by available tag keys",
                "filterValueTitle": "Filter by enabled tag keys",
                "filterValueText": "Remove your filter to see all enabled tag keys",
                "filterOptionsText": "Remove your filter to see all available tag keys",
                "initialValue": [],
                "clearedValue": []
            },
            {
                "label": "Amazon Web Services tags",
                "name": "api.settings.aws.title",
                "variant": "h2",
                "component": "plain-text"
            },
            {
                "component": "dual-list-select",
                "name": "api.settings.tag-management.aws.enabled",
                "options": [],
                "leftTitle": "Available tags",
                "rightTitle": "Tags for reporting",
                "noValueTitle": "No enabled tag keys",
                "noOptionsTitle": "No available tag keys",
                "filterOptionsTitle": "Filter by available tag keys",
                "filterValueTitle": "Filter by enabled tag keys",
                "filterValueText": "Remove your filter to see all enabled tag keys",
                "filterOptionsText": "Remove your filter to see all available tag keys",
                "initialValue": [],
                "clearedValue": []
            },
            {
                "label": "Azure tags",
                "name": "api.settings.azure.title",
                "variant": "h2",
                "component": "plain-text"
            },
            {
                "component": "dual-list-select",
                "name": "api.settings.tag-management.azure.enabled",
                "options": [],
                "leftTitle": "Available tags",
                "rightTitle": "Tags for reporting",
                "noValueTitle": "No enabled tag keys",
                "noOptionsTitle": "No available tag keys",
                "filterOptionsTitle": "Filter by available tag keys",
                "filterValueTitle": "Filter by enabled tag keys",
                "filterValueText": "Remove your filter to see all enabled tag keys",
                "filterOptionsText": "Remove your filter to see all available tag keys",
                "initialValue": [],
                "clearedValue": []
            },
            {
                "label": "Google Cloud Platform tags",
                "name": "api.settings.gcp.title",
                "variant": "h2",
                "component": "plain-text"
            },
            {
                "component": "dual-list-select",
                "name": "api.settings.tag-management.gcp.enabled",
                "options": [],
                "leftTitle": "Available tags",
                "rightTitle": "Tags for reporting",
                "noValueTitle": "No enabled tag keys",
                "noOptionsTitle": "No available tag keys",
                "filterOptionsTitle": "Filter by available tag keys",
                "filterValueTitle": "Filter by enabled tag keys",
                "filterValueText": "Remove your filter to see all enabled tag keys",
                "filterOptionsText": "Remove your filter to see all available tag keys",
                "initialValue": [],
                "clearedValue": []
            }
        ],
        "component": "sub-form"
    },
    {
        "title": "Currency",
        "fields": [
            {
                "label": "Select the prefered currency to view Cost Information in",
                "name": "api.settings.openshift.title",
                "component": "plain-text"
            },
            {
                "component": "select",
                "name": "api.settings.currency",
                "label": "Currency",
                "options": [
                    {
                        "value": "AUD",
                        "label": "AUD (AU$) - Australian Dollar"
                    },
                    {
                        "value": "CAD",
                        "label": "CAD (C$) - Canadian Dollar"
                    },
                    {
                        "value": "CHF",
                        "label": "CHF (CHF) - Swiss Franc"
                    },
                    {
                        "value": "CNY",
                        "label": "CNY (CN¥) - Chinese Yuan"
                    },
                    {
                        "value": "DKK",
                        "label": "DKK (Dkr) - Danish Krone"
                    },
                    {
                        "value": "EUR",
                        "label": "EUR (€) - Euro"
                    },
                    {
                        "value": "GBP",
                        "label": "GBP (£) - British Pound"
                    },
                    {
                        "value": "HKD",
                        "label": "HKD (HK$) - Hong Kong Dollar"
                    },
                    {
                        "value": "JPY",
                        "label": "JPY (¥) - Japanese Yen"
                    },
                    {
                        "value": "NOK",
                        "label": "NOK (Nkr) - Norwegian Krone"
                    },
                    {
                        "value": "NZD",
                        "label": "NZD (NZ$) - New Zealand Dollar"
                    },
                    {
                        "value": "SEK",
                        "label": "SEK (Skr) - Swedish Krona"
                    },
                    {
                        "value": "SGD",
                        "label": "SGD (S$) - Singapore Dollar"
                    },
                    {
                        "value": "USD",
                        "label": "USD ($) - United States Dollar"
                    },
                    {
                        "value": "ZAR",
                        "label": "ZAR (R) - South African Rand"
                    }
                ],
                "initialValue": "JPY"
            }
        ]
    }
]
```

### CODE REVIEW

File: koku/api/currency/view.py [hunk 1]
```diff
@@ -46,7 +38,7 @@ def get_currency(request):
 
     """
     try:
-        data = load_currencies(CURRENCY_FILE_NAME)
+        data = load_currencies_from_file()
```
Comment (boaz0) [inline koku/api/currency/view.py:41]: @dccurtis we should also add to the response the selected currency from settings.

@ddonahue007 @dlabrecq - is that right?

File: koku/currency/common.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+CURRENCY_FILE_NAME = f"{os.path.dirname(os.path.realpath(__file__))}/specs/currencies.json"
+
+
+def load_currencies_from_file(file_path=CURRENCY_FILE_NAME):
+    with open(file_path) as api_file:
+        data = json.load(api_file)
+        return data
+
+
+def get_selected_currency_or_setup(schema):
```
Comment (boaz0) [inline koku/currency/common.py:23]: Maybe we can move this to a new middleware or to the koku user middelware to set the currency setting to the KOKU_DEFAULT_CURRENCY on account creation?

**Review by boaz0 — COMMENTED** [review]
@ddonahue007 @dlabrecq @dccurtis - my thoughts in regards to add "selected currency" in the  `/currency/` endpoint response.

**Comment by boaz0** [comment]
BTW, I tested this with the settings frontend and it worked well but feel free to run your own checks and testings.

File: koku/api/settings/view.py [hunk 3]
```diff
@@ -14,9 +15,13 @@
 
 from api.common.permissions.settings_access import SettingsAccessPermission
 from api.settings.tag_management import TagManagementSettings
+from currency.settings import CurrencySettings
 
 LOG = logging.getLogger(__name__)
-SETTINGS_GENERATORS = {"tag-management": TagManagementSettings}
+if settings.DEVELOPMENT:
```
Comment (maskarb) [inline koku/api/settings/view.py:21]: Unleash will _not_ work here, at least, not the way in which we would want. This variable is initialized during Pod startup and then never touched again. For this logic to work with Unleash, we would need to flip a flag in Unleash, then restart the pod.

Every time the `settings` api is accessed, we need to check if things are enabled in Unleash.

The logic probably needs to live in the `_build_settings` method in this file below.

**Review by ddonahue007 — APPROVED** [review]
LGTM

**Pull request merged**